src/object.ts

```diff
@@ -79,10 +78,36 @@ function unknown(ctx: ObjectSchema<any, any, any>, value: any) {
 
 const defaultSort = sortByKeyOrder([]);
 
+type MatchShape<T extends Maybe<AnyObject>, C = any> = {
+  [field in keyof T]: ISchema<T[field], C> | Reference;
+};
+
+// Equals true if T is an object without known or explicitly-defined keys (e.g. AnyObject)
+type IsAnonymousObject<T> = string | number extends keyof T ? true : false;
+
+// Decides between an explicitly defined target type versus resolving the shape from a schema spec
+type TypeFromTarget<Target, S extends ObjectShape> = true extends IsAnonymousObject<Target>
+  ? MakePartial<{ 
+    [K in keyof S]: 
+      // If Reference, choose reference type
+      S[K] extends Reference<infer R> 
+      ? R 
+      // If schema, resolve the schema
+      : S[K] extends ISchema<any> 
+        ? ResolveStrip<S[K]> 
+        : S[K] 
+  }>
+  : Target;
+
+export function create(): ObjectSchema<{}, AnyObject, {}>;
+
 export function create<
+  TIn extends AnyObject,
   C extends Maybe<AnyObject> = AnyObject,
-  S extends ObjectShape = {},
->(spec?: S) {
+  S extends MatchShape<TIn, C> = MatchShape<TIn, C>
```

Comment:
A challenge here is that it's no longer possible to specify the context without explicitly defining TIn, e.g  `create<Context>({ ... })`

Reply:
Is `Context` ever explicitly defined via `object()`?  I thought it was only passed down/inherited from other methods.  I couldn't find any examples of `object<...>` in the codebase or documentation, so I wasn't aware this was something supported.

Do you have an example to illustrate?

Reply:
Context is always explicitly provided if used. It defines the arbitrary extra data made available to tests and transforms. E.g. 

```
object<{ api: APIClient }>().test ((value, schema)=> schema.options.context.api)
```

Reply:
Okay, I see now.  Its defined ahead of time, but the value is passed in during validation calls via `ValidateOptions`.

I think unfortunately this means fundamentally the first generic arg can't be the schema's type.

What are your thoughts if I move the explicit type ability to `.shape()` method instead?

Reply:
@jquense Did you see my question above?